osquery/tables/system/windows/objects.cpp

```diff
@@ -0,0 +1,214 @@
+// converted to UTF-8 encoded std::strings when communicating with the
+// rest of osquery
+//
+typedef std::pair<std::wstring, std::wstring> obj_name_type_pair;
+
+// helper to convert a std::wstring to an integer
+// std::stoi can throw, and std::stol doesn't support std::wstring
+static int safe_wstr_to_int(std::wstring str) {
+  try {
+    return std::stoi(str);
+  }
+  catch (const std::out_of_range&) {
+    return 0;
+  }
+  catch (const std::invalid_argument&) {
+    return 0;
+  }
+}
+
+// enumerate all objects in the Windows object namespace
+// does not provide support for recursion
+//
+// example at:
+//    http://pastebin.com/embed_js/zhmJTffK
+//    https://randomsourcecode.wordpress.com/2015/03/14/enumerating-deviceobjects-from-user-mode/
+//    https://msdn.microsoft.com/en-us/library/bb470238(v=vs.85).aspx
+//
+std::vector<obj_name_type_pair> EnumerateObjectNamespace(std::wstring directory) {
+  std::vector<obj_name_type_pair> objects;
+
+  // look up addresses of NtQueryDirectoryObject and
+  // NtQuerySymbolicLinkObject.  Both are exported from ntdll
+  //
+  // NtQueryDirectoryObject is documented on MSDN, there is no
+  // associated header or import library
+  NTQUERYDIRECTORYOBJECT NtQueryDirectoryObject =
+      (NTQUERYDIRECTORYOBJECT)GetProcAddress(GetModuleHandleA("ntdll"),
+                                             "NtQueryDirectoryObject");
+  if (NULL == NtQueryDirectoryObject) {
+    return objects;
+  }
+
+  // open the caller-provided root directory
+  // kdo will manage the resultant HANDLE, which is also used to query
+  // for object name and object type name pairs below
+  KObjHandle kdo;
+  if (!kdo.openDirObj(directory)) {
+    return objects;
+  }
+
+  // iterator index is incremented by NtQueryDirectoryObject
+  for (DWORD index = 0;;) {
+    BYTE rgDirObjInfoBuffer[1024 * 8] = {0};
+    POBJDIR_INFORMATION pObjDirInfo = (POBJDIR_INFORMATION)rgDirObjInfoBuffer;
+
+    //  get the name and type of the index'th object in the directory
+    NTSTATUS ntStatus = NtQueryDirectoryObject(kdo.getAsHandle(),
+                                               pObjDirInfo,
+                                               sizeof(rgDirObjInfoBuffer),
+                                               TRUE,
+                                               FALSE,
+                                               &index,
+                                               NULL);
+    if (STATUS_SUCCESS != ntStatus) {
+      break;
+    }
+
+    obj_name_type_pair object;
+    object.first = (pObjDirInfo->ObjectName.Buffer);
+    object.second = (pObjDirInfo->ObjectTypeName.Buffer);
+
+    objects.push_back(object);
+  }
+
+  return objects;
+}
+
+// enumerate all objects in a given windows terminal services session
+//
+// objects are found in the windows object directory
+// "\Sessions\BNOLINKS\<sessionnum>"
+//
+std::vector<obj_name_type_pair>
+EnumerateBaseNamedObjectsLinks(std::wstring session_num,
```

Comment:
done

Reply:
nit: lower on the function name here also, further let's change this to possibly be `Status enumerateBaseNamedObjectsLinks(std::wstring session_num, std::wstring object_type, std::vector<obj_name_type_pai> results) {` to avoid returning a potentially large amount of data.

Reply:
nit: lower on the function name here also, further let's change this to possibly be `Status enumerateBaseNamedObjectsLinks(std::wstring session_num, std::wstring object_type, std::vector<obj_name_type_pai> results) {` to avoid returning a potentially large amount of data.